### PR TITLE
[fsdp] fix: position_ids in qwen-vl

### PR DESCRIPTION
**Pull request by ShareLer**
### Checklist Before Starting

- [x] Searched for similar PR(s).
- [x] Checked PR Title format
  - [ ] In format of: [modules] type: Title
  - [ ] modules are in `fsdp, megatron, sglang, vllm, rollout, trainer, tests, training_utils, recipe, hardware, deployment, ray, worker, single_controller, misc, perf, model, algo, env, tool, ckpt`
  - [ ] type is in `feat, fix, doc, refactor, chore`
  - [ ] can involve multiple modules, seperated by `,` or space, like `[megatron, fsdp] feat: xxx`

### What does this PR do?

Fix two issues releated to position_ids for qwen2_VL/qwen2.5_VL:

(1) Create processor with use_fast=True lead to use `Qwen2VLImageProcessorFast`, however, when determining whether to handle 3D position ids, the Qwen2VLImageProcessor was still used. 
(2) And 3D position is not considered in ulysses_pad.

### High-Level Design

> Demonstrate the high-level design if this PR is complex.

### Specific Changes

> List the specific changes.

### API

> Demonstrate how the API changes if any.

### Usage Example

> Provide usage example(s) for easier usage.

```python
# Add code snippet or script demonstrating how to use this 
```

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluatuion results, etc.

### Additional Info.

- **Issue Number**: Fixes issue # or discussion # if any.
- **Training**: [Note which backend this PR will affect: FSDP, Megatron, both, or none]
- **Inference**: [Note which backend this PR will affect: vLLM, SGLang, both, or none]

### Checklist Before Submitting

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting).
- [ ] Add `[BREAKING]` to the PR title if it breaks any API.
- [ ] Update the documentation about your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs).
- [ ] New CI unit test(s) are added to cover the code path.
- [ ] Rely on existing unit tests on CI that covers the code path.

### CODE REVIEW

**Comment by TheEighthDay** [comment]
hi, i get error in   

https://github.com/volcengine/verl/blob/main/verl/workers/fsdp_workers.py#L1284

eos_mask_idx = torch.argmax(position_ids * attention_mask, dim=-1)  # (bsz,)

**Comment by ShareLer** [comment]
> hi, i get error in
> 
> https://github.com/volcengine/verl/blob/main/verl/workers/fsdp_workers.py#L1284
> 
> eos_mask_idx = torch.argmax(position_ids * attention_mask, dim=-1) # (bsz,)

Thanks. I have fixed some hidden issues. You can try the new code again.

**Comment by TheEighthDay** [comment]
tmp_position_ids = position_ids[:,0,:] 
#maybe multiple max value for visual token in the first dim of m-ROPE, but argamx only return the first value index
maybe:
tmp_position_ids = torch.clip(torch.cumsum(attention_mask, dim=-1) - 1, min=0, max=None)

tmp_position_ids will be used for:
eos_mask_idx = torch.argmax(tmp_position_ids * attention_mask, dim=-1)  # (bsz,)

**Comment by hiyouga** [comment]
> tmp_position_ids = position_ids[:,0,:] #maybe multiple max value for visual token in the first dim of m-ROPE, but argamx only return the first value index maybe: tmp_position_ids = torch.clip(torch.cumsum(attention_mask, dim=-1) - 1, min=0, max=None)
> 
> tmp_position_ids will be used for: eos_mask_idx = torch.argmax(tmp_position_ids * attention_mask, dim=-1) # (bsz,)

agreed

**Comment by ShareLer** [comment]
> tmp_position_ids = position_ids[:,0,:] #maybe multiple max value for visual token in the first dim of m-ROPE, but argamx only return the first value index maybe: tmp_position_ids = torch.clip(torch.cumsum(attention_mask, dim=-1) - 1, min=0, max=None)
> 
> tmp_position_ids will be used for: eos_mask_idx = torch.argmax(tmp_position_ids * attention_mask, dim=-1) # (bsz,)

After passing through apply_chat_template, the image_token is surrounded by <|vision_start|> and <|vision_end|>. Therefore, in whole position_ids in first dim, there will always be a value larger than the postion_id corresponding to image_tokend even if the picture is after the text.

Example:
input_ids: TTTTSVVVVVVVVVE
position_ids[0]: 012345555555556
 (T is text_token, V is image/vidio token, S is <|vision_start|> and E is <|vision_end|>)

But your method will indeed be more secure.

**Comment by hiyouga** [comment]
I'll rerun the failed job and merge this PR if it passed

**Review by hiyouga — APPROVED** [review]
ready for merge @vermouth1992

**Comment by ShareLer** [comment]
> I'll rerun the failed job and merge this PR if it passed

Thanks for your review and revision.
